user/super/com/google/gwt/emul/java/lang/String.java

```diff
@@ -755,6 +760,82 @@ public String trim() {
     return start > 0 || end < length ? substring(start, end) : this;
   }
 
+  public String strip() {
+    int length = length();
+    int start = getLeadingWhitespaceLength();
+    if (start == length) {
+      return "";
+    }
+    return substring(start, length - getTrailingWhitespaceLength());
+  }
+
+  public String stripLeading() {
+    return substring(getLeadingWhitespaceLength());
+  }
+
+  public String stripTrailing() {
+    return substring(0, length() - getTrailingWhitespaceLength());
+  }
+
+  public boolean isBlank() {
+    return length() == getLeadingWhitespaceLength();
+  }
+
+  public Stream<String> lines() {
+    return StreamSupport.stream(new LinesSpliterator(this), false);
+  }
+
+  public String repeat(int count) {
+    if (count < 0) {
+      throw new IllegalArgumentException("count is negative: " + count);
+    }
+    return asNativeString().repeat(count);
+  }
+
+  private int getLeadingWhitespaceLength() {
+    int length = length();
+    for (int i = 0; i < length; i++) {
+      if (!Character.isWhitespace(charAt(i))) {
+        return i;
+      }
+    }
+    return length;
+  }
+
+  private int getTrailingWhitespaceLength() {
+    int length = length();
+    for (int i = length - 1; i >= 0; i--) {
+      if (!Character.isWhitespace(charAt(i))) {
+        return length - 1 - i;
+      }
+    }
+    return length;
+  }
+
+  private static class LinesSpliterator extends Spliterators.AbstractSpliterator<String> {
+    private int processed = 0;
+    private String content;
+
+    private LinesSpliterator(String content) {
+      super(Long.MAX_VALUE, Spliterator.IMMUTABLE | Spliterator.ORDERED);
+      this.content = content;
+    }
+
+    @Override
+    public boolean tryAdvance(Consumer<? super String> action) {
+      int rPosition = content.indexOf('\r', processed);
+      int nPosition = content.indexOf('\n', processed);
+      int lineEnd = nPosition == -1 ? rPosition : Math.min(nPosition, rPosition);
+      action.accept(lineEnd == -1 ? content.substring(processed)
+              : content.substring(processed, lineEnd));
+      processed = lineEnd + 1;
+      if (nPosition == rPosition + 1 && rPosition != -1) {
+        processed++;
+      }
+      return lineEnd != -1 && processed < content.length();
```

Comment:
if you expand the ternary into an if, you can avoid running the other 4 lines of code at all when lineEnd == -1, roughly:
```suggestion
      if (lineEnd == -1) {
        action.accept(content.substring(processed));
        processed = content.length();
        return false;
      }
      action.accept(content.substring(processed, lineEnd));
      processed = lineEnd + 1;
      if (nPosition == rPosition + 1 && rPosition != -1) {
        processed++;
      }
      return processed < content.length();
```

I'm not sure if this makes any meaningful difference, but it was at least a bit easier to read... If we had a loop here it could matter, but all the "hard" work is already done by this point.

---

That said, do we need to do both indexOf()s unconditionally? Can we not instead scan from `processed` to find `'\r'` - if success test next char for `'\n'`, if fail re-scan from `processed`?

Reply:
>  Can we not instead scan from processed to find '\r' - if success test next char for '\n', if fail re-scan from processed?

We have to be careful about `a\nb\r\n` where 'scanning for `\r` first' skips the first `\n`, but your point is valid -- currently for the string `\n\n\n\n\n\n` we'd check `O(length^2)` times if a character is `\r`, which is terrible. If we don't recompute the positions that are still valid from last iteration we can only check each character once for `\r` and once for `\n` which is optimal.